File: question/leetcode/leetcode0019-Remove-Nth-Node-From-End-of-List/leetcode0019.go
package leetcode

import "clarencedu/algorithm-go/structure"

type ListNode = structure.ListNode

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	targetNode := findNthFromEnd(dummy, n+1)
	targetNode.Next = targetNode.Next.Next
	return dummy.Next
}

func findNthFromEnd(head *ListNode, n int) *ListNode {
	if head == nil {
		return nil
	}
	fast := head
	for i := 0; i < n; i++ {
		fast = fast.Next
	}
	slow := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	return slow
}

func removeNthFromEndII(head *ListNode, n int) *ListNode {
	dummy := &ListNode{0, head}
	cur := dummy
	length := 0
	for ; head != nil; head = head.Next {
		length++
	}

	for i := 0; i < length-n; i++ {
		cur = cur.Next
	}
	cur.Next = cur.Next.Next

	return dummy.Next
}

File: question/leetcode/leetcode0019-Remove-Nth-Node-From-End-of-List/leetcode0019.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode targetNode = findNthFromEnd(dummy, n + 1);
        targetNode.next = targetNode.next.next;

        return dummy.next;
    }

    public ListNode findNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        for (int i = 0; i < n; i++) {
            if (fast == null) {
                return null;
            }
            fast = fast.next;
        }
        ListNode slow = head;
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}